DevWs for WiDom srl containing containing WiDom Smart Dry Contact

GiuseppeWidom opened the pull request.

GiuseppeWidom committed:
DevWs for WiDom srl containing containing WiDom Smart Dry Contact

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
```
Please add in definition() 
ocfDeviceType: "oic.d.switch", mnmn: "SmartThings", **vid: "generic-switch"**
(If your device reports power then  **vid: "generic-switch-power**",  if it reports power and energy then **vid: "generic-switch-power"**)

GiuseppeWidom replied:
Fixed with:
definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl", ocfDeviceType: "oic.d.switch", mnmn: "SmartThings", vid: "generic-switch") {

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+		capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"
```
Please add deviceJoinName "WiDom Switch" in this fingerprint and add Raw Description in a comment after the fingerprint so it would look like:
fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName "WiDom Switch" // Raw Description (paste the raw description here)

(Raw description can be found at the Smartthings graphapi in device details)

KKlimczukS replied:
What about this remark ?

KKlimczukS replied:
Please fix this one too (according to my previous remarks).

GiuseppeWidom replied:
@ KKlimczukS
I fixed the Raw description the 1st April 2021 after request of @PKacprowiczS

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+        fingerprint mfr: "0149", prod: "1214", model: "0900"
+    }
+    
+   	tiles{
```
Please remove the whole tiles section since that's not supported in the SmartThings app.

GiuseppeWidom replied:
Removed all tiles

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+		//capability "Actuator"
```
Is this the complete list of capabilities ? What about capability "Health Check"?

KKlimczukS replied:
please also add: 
capability "Actuator"

GiuseppeWidom replied:
Fixed with:
		capability "Actuator"
		capability "Switch"
		capability "Configuration"
		capability "Health Check"

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
```
Please reformat the whole code and fix indentation (change 4 spaces to 1 tab).

GiuseppeWidom replied:
fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+    def cmds = []
+    logging("Executing updated()","info")
+
+    if (device.currentValue("numberOfButtons") != 6) { sendEvent(name: "numberOfButtons", value: 6) }
```
Since this DTH was based on fibaro-double-switch-2-zw5.groovy, please remove all the code that is related to Fibaro switch.

GiuseppeWidom replied:
Fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+    [0x5E: 2, 0x86: 2, 0x72: 2, 0x59: 2, 0x98: 1, 0x25: 1, 0x5A: 1, 0x85: 2, 0x70: 1, 0x8E: 2, 0x9F: 1, 0x6C: 1] 
+}
+
+private parameterMap() {[
```
Are all those parameters related to scenes and automations necessary here ?

GiuseppeWidom replied:
The parameters are necessary for change the device behavior(for example number of click on external button in order to turn ON/OFF the load).

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+    encap(multiEncap(cmd, ep))
+}
+
+private encap(List encapList) {
```
Please, could You point where do You use this method with a list?

GiuseppeWidom replied:
Method removed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,548 @@
+    encap(encapList[0], encapList[1])
+}
+
+private encap(Map encapMap) {
```
Please, could You point where do You use this method with a map?

GiuseppeWidom replied:
Method removed

GiuseppeWidom committed:
Added in definition() "ocfDeviceType"
Added deviceJoinName "WiDom Dry Contact" in this fingerprint and added Raw description in a comment
Removed the whole tiles section
Reformated the whole code and fixed indentation (change 4 spaces to 1 tab).
Removed useless code.

GiuseppeWidom commented:
Hello,

I update the code following your istructions.





Giuseppe Tiddia
--------------- 
R&D Engineer
WiDom Srl
Tel.: 0702057916 
e-mail: giuseppe.tiddia@widom.it

Via G. Mameli, 88 - 09123 Cagliari - IT
www.widom.it 





Da: "Konrad K" notifications@github.com
A: "SmartThingsCommunity/SmartThingsPublic" SmartThingsPublic@noreply.github.com
Cc: "GiuseppeWidom" giuseppe.tiddia@widom.it,"Author" author@noreply.github.com
Data: Tue, 12 Jan 2021 09:15:30 -0800
Oggetto: Re: [SmartThingsCommunity/SmartThingsPublic] DevWs for WiDom srl containing containing WiDom Smart Dry Contact (#52683)



@KKlimczukS requested changes on this pull request.



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> @@ -0,0 +1,548 @@
+/**
+ *  Widom Smart DRY contact
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License. You may obtain a copy of the License at:
+ *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {

Please add in definition()

ocfDeviceType: "oic.d.switch", mnmn: "SmartThings", vid: "generic-switch"

(If your device reports power then  vid: "generic-switch-power",  if it reports power and energy then vid: "generic-switch-power")



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> + *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+		//capability "Actuator"
+		capability "Switch"
+		capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"

Please add deviceJoinName "WiDom Switch" in this fingerprint and add Raw Description in a comment after the fingerprint so it would look like:

fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName "WiDom Switch" // Raw Description 
(Raw description can be found at the Smartthings graphapi in device details)



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> + *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+		//capability "Actuator"
+		capability "Switch"
+		capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"
+    }
+    
+   	tiles{

Please remove the whole tiles section since that's not supported in the SmartThings app.



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +/**
+ *  Widom Smart DRY contact
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License. You may obtain a copy of the License at:
+ *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+		//capability "Actuator"

Is this the complete list of capabilities ? What about capability "Health Check"?



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> @@ -0,0 +1,548 @@
+/**
+ *  Widom Smart DRY contact
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License. You may obtain a copy of the License at:
+ *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {

Please reformat the whole code and fix indentation (change 4 spaces to 1 tab).



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +def on() {
+    encap(zwave.basicV1.basicSet(value: 255))
+}
+
+def off() {
+    encap(zwave.basicV1.basicSet(value: 0))
+}
+
+
+//Configuration and synchronization
+def updated() {
+    if ( state.lastUpdated && (now() - state.lastUpdated) < 500 ) return
+    def cmds = []
+    logging("Executing updated()","info")
+
+    if (device.currentValue("numberOfButtons") != 6) { sendEvent(name: "numberOfButtons", value: 6) }

Since this DTH was based on fibaro-double-switch-2-zw5.groovy, please remove all the code that is related to Fibaro switch.



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +    else {
+        return []
+    }
+}
+
+
+/*
+##########################
+## Device Configuration ##
+##########################
+*/
+private Map cmdVersions() {
+    [0x5E: 2, 0x86: 2, 0x72: 2, 0x59: 2, 0x98: 1, 0x25: 1, 0x5A: 1, 0x85: 2, 0x70: 1, 0x8E: 2, 0x9F: 1, 0x6C: 1] 
+}
+
+private parameterMap() {[

Are all those parameters related to scenes and automations necessary here ?



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +
+private crcEncap(physicalgraph.zwave.Command cmd) {
+    logging("encapsulating command using CRC16 Encapsulation, command: $cmd","info")
+    zwave.crc16EncapV1.crc16Encap().encapsulate(cmd).format()
+}
+
+private multiEncap(physicalgraph.zwave.Command cmd, Integer ep) {
+    logging("encapsulating command using MultiChannel Encapsulation, ep: $ep command: $cmd","info")
+    zwave.multiChannelV3.multiChannelCmdEncap(destinationEndPoint:ep).encapsulate(cmd)
+}
+
+private encap(physicalgraph.zwave.Command cmd, Integer ep) {
+    encap(multiEncap(cmd, ep))
+}
+
+private encap(List encapList) {

Please, could You point where do You use this method with a list?



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +}
+
+private multiEncap(physicalgraph.zwave.Command cmd, Integer ep) {
+    logging("encapsulating command using MultiChannel Encapsulation, ep: $ep command: $cmd","info")
+    zwave.multiChannelV3.multiChannelCmdEncap(destinationEndPoint:ep).encapsulate(cmd)
+}
+
+private encap(physicalgraph.zwave.Command cmd, Integer ep) {
+    encap(multiEncap(cmd, ep))
+}
+
+private encap(List encapList) {
+    encap(encapList[0], encapList[1])
+}
+
+private encap(Map encapMap) {

Please, could You point where do You use this method with a map?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

GiuseppeWidom committed:
Removed useless code

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,353 @@
+	encap(multiEncap(cmd, ep))
+}
+
+//private encap(List encapList) {
```
I guess below encap() methods can be removed now, right ?

GiuseppeWidom replied:
Removed encap method

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,353 @@
+		input ( name: "logging", title: "Logging", type: "boolean", required: false )
+	}
+}
+
```
Please remove unnecessary double line breaks, one line break is sufficient (applies to a whole file).

GiuseppeWidom replied:
fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,353 @@
+	}
+    
+
+	if ( syncNeeded ) {
```
Spaces inside brackets of an if statement / with one condition only - are not necessary.

GiuseppeWidom replied:
fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,353 @@
+	size.times {
+		result = result.plus(0, (value & 0xFF) as Short)
+		value = (value >> 8)
+		}
```
Please fix indentation.

GiuseppeWidom replied:
fixed

KKlimczukS reviewed (changes requested):
Please refer also to all remarks from previous review.

GiuseppeWidom committed:
Deleted space inside if statements with one condition
Deleted double empty lines and useless lines
Indentation fix

GiuseppeWidom commented:
Hello,
   I update the code following your istructions.

   Best regards
   Giuseppe Tiddia
   ---------------
   R&D Engineer
   WiDom Srl
   Tel.: 0702057916
   e-mail: giuseppe.tiddia@widom.it
   Via G. Mameli, 88 - 09123 Cagliari - IT
   www.widom.it

   Da "Konrad K" notifications@github.com
   A "SmartThingsCommunity/SmartThingsPublic" SmartThingsPublic@noreply.github.com
   Cc "GiuseppeWidom" giuseppe.tiddia@widom.it,"Author" author@noreply.github.com
   Data Thu, 21 Jan 2021 07:00:23 -0800
   Oggetto Re: [SmartThingsCommunity/SmartThingsPublic] DevWs for WiDom srl containing containing WiDom Smart Dry Contact (#52683)

   @KKlimczukS requested changes on this pull request.

   Please refer also to all remarks from previous review.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +
+private crcEncap(physicalgraph.zwave.Command cmd) {
+       logging("encapsulating command using CRC16 Encapsulation, command: $cmd","info")
+       zwave.crc16EncapV1.crc16Encap().encapsulate(cmd).format()
+}
+
+private multiEncap(physicalgraph.zwave.Command cmd, Integer ep) {
+       logging("encapsulating command using MultiChannel Encapsulation, ep: $ep command: $cmd","info")
+       zwave.multiChannelV3.multiChannelCmdEncap(destinationEndPoint:ep).encapsulate(cmd)
+}
+
+private encap(physicalgraph.zwave.Command cmd, Integer ep) {
+       encap(multiEncap(cmd, ep))
+}
+
+//private encap(List encapList) {

   I guess below encap() methods can be removed now, right ?
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +
+                       input (
+                                       name: it.key,
+                                       title: null,
+                                       //description: "Default: $it.def" ,
+                                       type: it.type,
+                                       options: it.options,
+                                       //range: (it.min != null && it.max != null) ? "${it.min}..${it.max}" : null,
+                                       defaultValue: it.def,
+                                       required: false
+                       )
+               }
+               input ( name: "logging", title: "Logging", type: "boolean", required: false )
+       }
+}
+

   Please remove unnecessary double line breaks, one line break is sufficient (applies to a whole file).
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +     boolean syncNeededGroup = false
+       Integer settingValue = null
+       parameterMap().each {
+               if(settings."$it.key" != null) {
+                       settingValue = settings."$it.key" as Integer
+                       if (state."$it.key" == null) { state."$it.key" = [value: null, state: "synced"] }
+                       if (state."$it.key".value != settingValue || state."$it.key".state != "synced" ) {
+                               state."$it.key".value = settingValue
+                               state."$it.key".state = "notSynced"
+                               syncNeeded = true
+                       }
+               }
+       }
+
+
+       if ( syncNeeded ) {

   Spaces inside brackets of an if statement / with one condition only - are not necessary.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +}
+
+private encapSequence(cmds, Integer delay=250) {
+       delayBetween(cmds.collect{ encap(it) }, delay)
+}
+
+private encapSequence(cmds, Integer delay, Integer ep) {
+       delayBetween(cmds.collect{ encap(it, ep) }, delay)
+}
+
+private List intToParam(Long value, Integer size = 1) {
+       def result = []
+       size.times {
+               result = result.plus(0, (value & 0xFF) as Short)
+               value = (value >> 8)
+               }

   Please fix indentation.

   —
   You are receiving this because you authored the thread.
   Reply to this email directly, view it on GitHub, or unsubscribe.

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,339 @@
+def zwaveEvent(physicalgraph.zwave.commands.securityv1.SecurityMessageEncapsulation cmd) {
+	def encapsulatedCommand = cmd.encapsulatedCommand(cmdVersions())
+	if (encapsulatedCommand) {
+    	logging("Parsed SecurityMessageEncapsulation into: ${encapsulatedCommand}")
```
Please fix this indentation (4 spaces -> 1 tab)

GiuseppeWidom replied:
Fixed

GiuseppeWidom committed:
added actuator capability, indentation fix

GiuseppeWidom commented:
I made the requested changes.RegardsGiuseppe Tiddia--------------- R&D EngineerWiDom SrlTel.: 0702057916 e-mail: giuseppe.tiddia@widom.itVia G. Mameli, 88 - 09123 Cagliari - ITwww.widom.it 


Da: "Konrad K" notifications@github.com
A: "SmartThingsCommunity/SmartThingsPublic" SmartThingsPublic@noreply.github.com
Cc: "GiuseppeWidom" giuseppe.tiddia@widom.it,"Author" author@noreply.github.com
Data: Fri, 29 Jan 2021 03:28:38 -0800
Oggetto: Re: [SmartThingsCommunity/SmartThingsPublic] DevWs for WiDom srl containing containing WiDom Smart Dry Contact (#52683)


@KKlimczukS requested changes on this pull request.



In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +		deviceId = description.split(", ")[0]
+		deviceId = deviceId.split(":")[1]
+		logging("deviceId- ${deviceId}")
+		deviceIdEvent(deviceId, true, true)
+		def cmd = zwave.parse(description, cmdVersions())
+		if (cmd) {
+			logging("Parsed: ${cmd}")
+			zwaveEvent(cmd)
+		}
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.securityv1.SecurityMessageEncapsulation cmd) {
+	def encapsulatedCommand = cmd.encapsulatedCommand(cmdVersions())
+	if (encapsulatedCommand) {
+    	logging("Parsed SecurityMessageEncapsulation into: ${encapsulatedCommand}")

Please fix this indentation (4 spaces -> 1 tab)

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+
+private parameterMap() {[
+		[key: "numClickLoad", num: 1, size: 1, type: "number", min: 0, max: 7, def: 7, title: "Numbers of clicks to controlthe loads",
+			descr: "Define which sequences of clicks control the load (see device manual)."],     
```
Please remove those unnecessary 4 spaces at the end of the line.

GiuseppeWidom replied:
fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+	logging("Parsing: ${description}")
+	if (description.startsWith("Err 106")) {
+	result = createEvent(
+				descriptionText: "Failed to complete the network security key exchange. If you are unable to receive data from it, you must remove it from your network and add it again.",
```
Could You fix the indentation here ? (it seems that 3 tabs are enough here)

GiuseppeWidom replied:
fixed

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+		[key: "threeClickScene", num: 22, size: 2, type: "number",min: 0, max: 255, def: 0, title: "Three Clicks Scene ActivationSet",
+			descr: "Defines the Scene Activation Set value sent to the Lifeline group with 1 Clicks on the external switch"],
+		[key: "startUpStatus", num: 60, size: 1, type: "enum", options: [
+				1: "ON",
```
Could You fix the indentation here ? (it seems that 3 tabs are enough here)

GiuseppeWidom replied:
fixed

GiuseppeWidom commented:
Hello,

   I have already fix this remark:
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+               //capability "Actuator"
+               capability "Switch"
+               capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"

   From Smartthings IDE I have committed this:
   definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl", ocfDeviceType: "oic.d.switch", mnmn: "SmartThings", vid: "generic-switch") {
   capability "Actuator"
   capability "Switch"
   capability "Configuration"
   capability "Health Check"
   fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName: "WiDom Dry Contact" // WiDom Dry Contact allow to turn On/Off a load with indipendent power supply.
   }
   Can you see my changes?
   Which is the procedure to commit the fixes for the my pull request? Commit my changes from the Smartthings IDE is enough?
   Giuseppe Tiddia
   ---------------
   R&D Engineer
   WiDom Srl
   Tel.: 0702057916
   e-mail: giuseppe.tiddia@widom.it
   Via G. Mameli, 88 - 09123 Cagliari - IT
   www.widom.it
   Da "Konrad K" notifications@github.com
   A "SmartThingsCommunity/SmartThingsPublic" SmartThingsPublic@noreply.github.com
   Cc "GiuseppeWidom" giuseppe.tiddia@widom.it,"Author" author@noreply.github.com
   Data Sun, 07 Feb 2021 23:24:00 -0800
   Oggetto Re: [SmartThingsCommunity/SmartThingsPublic] DevWs for WiDom srl containing containing WiDom Smart Dry Contact (#52683)

   @KKlimczukS commented on this pull request.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +     }
+       return result
+}
+
+/*
+##########################
+## Device Configuration ##
+##########################
+*/
+private Map cmdVersions() {
+       [0x5E: 2, 0x86: 2, 0x72: 2, 0x59: 2, 0x98: 1, 0x25: 1, 0x5A: 1, 0x85: 2, 0x70: 1, 0x8E: 2, 0x6C: 1]
+}
+
+private parameterMap() {[
+               [key: "numClickLoad", num: 1, size: 1, type: "number", min: 0, max: 7, def: 7, title: "Numbers of clicks to controlthe loads",
+                       descr: "Define which sequences of clicks control the load (see device manual)."],

   Please remove those unnecessary 4 spaces at the end of the line.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +     logging("SwitchBinaryReport received, value: ${cmd.value} ","info")
+       sendEvent([name: "switch", value: (cmd.value == 0 ) ? "off": "on"])
+}
+
+/*
+####################
+## Z-Wave Toolkit ##
+####################
+*/
+def parse(String description) {
+       def result = []
+       def deviceId = [];
+       logging("Parsing: ${description}")
+       if (description.startsWith("Err 106")) {
+       result = createEvent(
+                               descriptionText: "Failed to complete the network security key exchange. If you are unable to receive data from it, you must remove it from your network and add it again.",

   Could You fix the indentation here ? (it seems that 3 tabs are enough here)
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +             [key: "OffTimer", num: 10, size: 2, type: "number", def: 0, min: 0, max: 32000, title: " Timer to switch OFF the Relay",
+                       descr: "Defines the time after which the relay is switched OFF. Time unit is set by parameter 15(see device manual)"],
+               [key: "OnTimer", num: 11, size: 2, type: "number", def: 0, min: 0, max: 32000, title: " Timer to switch ON the Relay",
+                       descr: "Defines the time after which the relay is switched ON. Time unit is set by parameter 15(see device manual)"],
+               [key: "timerScale", num: 15, size: 1, type: "enum", options: [
+                       1: "Tenth of seconds",
+                       2: "Seconds",
+                       ], def: "1", title: "Timer scale", descr: "Defines the time unit used for parameters No.10 and No.11"],
+               [key: "oneClickScene", num: 20, size: 2, type: "number",min: 0, max: 255, def: 0, title: "One Click Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 1 Clickon the external switch"],
+               [key: "twoClickScene", num: 21, size: 2, type: "number",min: 0, max: 255, def: 0, title: "Two Clicks Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 2 Clickson the external switch"],
+               [key: "threeClickScene", num: 22, size: 2, type: "number",min: 0, max: 255, def: 0, title: "Three Clicks Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 1 Clicks on the external switch"],
+               [key: "startUpStatus", num: 60, size: 1, type: "enum", options: [
+                               1: "ON",

   Could You fix the indentation here ? (it seems that 3 tabs are enough here)
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> + *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+               //capability "Actuator"
+               capability "Switch"
+               capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"

   Please fix this one too (according to my previous remarks).

   —
   You are receiving this because you authored the thread.
   Reply to this email directly, view it on GitHub, or unsubscribe.

GiuseppeWidom committed:
Indentation Fix

GiuseppeWidom committed:
Indentation fix

GiuseppeWidom commented:
Hello,

   I made the requested changes, can you confirm that you see the changes of my last commit?

   Thank you
   Best regards
   Giuseppe Tiddia
   ---------------
   R&D Engineer
   WiDom Srl
   Tel.: 0702057916
   e-mail: giuseppe.tiddia@widom.it
   Via G. Mameli, 88 - 09123 Cagliari - IT
   www.widom.it
   Da "Konrad K" notifications@github.com
   A "SmartThingsCommunity/SmartThingsPublic" SmartThingsPublic@noreply.github.com
   Cc "GiuseppeWidom" giuseppe.tiddia@widom.it,"Author" author@noreply.github.com
   Data Sun, 07 Feb 2021 23:24:00 -0800
   Oggetto Re: [SmartThingsCommunity/SmartThingsPublic] DevWs for WiDom srl containing containing WiDom Smart Dry Contact (#52683)

   @KKlimczukS commented on this pull request.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +     }
+       return result
+}
+
+/*
+##########################
+## Device Configuration ##
+##########################
+*/
+private Map cmdVersions() {
+       [0x5E: 2, 0x86: 2, 0x72: 2, 0x59: 2, 0x98: 1, 0x25: 1, 0x5A: 1, 0x85: 2, 0x70: 1, 0x8E: 2, 0x6C: 1]
+}
+
+private parameterMap() {[
+               [key: "numClickLoad", num: 1, size: 1, type: "number", min: 0, max: 7, def: 7, title: "Numbers of clicks to controlthe loads",
+                       descr: "Define which sequences of clicks control the load (see device manual)."],

   Please remove those unnecessary 4 spaces at the end of the line.
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +     logging("SwitchBinaryReport received, value: ${cmd.value} ","info")
+       sendEvent([name: "switch", value: (cmd.value == 0 ) ? "off": "on"])
+}
+
+/*
+####################
+## Z-Wave Toolkit ##
+####################
+*/
+def parse(String description) {
+       def result = []
+       def deviceId = [];
+       logging("Parsing: ${description}")
+       if (description.startsWith("Err 106")) {
+       result = createEvent(
+                               descriptionText: "Failed to complete the network security key exchange. If you are unable to receive data from it, you must remove it from your network and add it again.",

   Could You fix the indentation here ? (it seems that 3 tabs are enough here)
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> +             [key: "OffTimer", num: 10, size: 2, type: "number", def: 0, min: 0, max: 32000, title: " Timer to switch OFF the Relay",
+                       descr: "Defines the time after which the relay is switched OFF. Time unit is set by parameter 15(see device manual)"],
+               [key: "OnTimer", num: 11, size: 2, type: "number", def: 0, min: 0, max: 32000, title: " Timer to switch ON the Relay",
+                       descr: "Defines the time after which the relay is switched ON. Time unit is set by parameter 15(see device manual)"],
+               [key: "timerScale", num: 15, size: 1, type: "enum", options: [
+                       1: "Tenth of seconds",
+                       2: "Seconds",
+                       ], def: "1", title: "Timer scale", descr: "Defines the time unit used for parameters No.10 and No.11"],
+               [key: "oneClickScene", num: 20, size: 2, type: "number",min: 0, max: 255, def: 0, title: "One Click Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 1 Clickon the external switch"],
+               [key: "twoClickScene", num: 21, size: 2, type: "number",min: 0, max: 255, def: 0, title: "Two Clicks Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 2 Clickson the external switch"],
+               [key: "threeClickScene", num: 22, size: 2, type: "number",min: 0, max: 255, def: 0, title: "Three Clicks Scene ActivationSet",
+                       descr: "Defines the Scene Activation Set value sent to the Lifeline group with 1 Clicks on the external switch"],
+               [key: "startUpStatus", num: 60, size: 1, type: "enum", options: [
+                               1: "ON",

   Could You fix the indentation here ? (it seems that 3 tabs are enough here)
     _________________________________________________________________

   In devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
> + *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+metadata {
+    definition (name: "WiDom Smart Dry Contact", namespace: "WiDomsrl", author: "WiDom srl") {
+               //capability "Actuator"
+               capability "Switch"
+               capability "Relay Switch"
+        capability "Configuration"
+
+        fingerprint mfr: "0149", prod: "1214", model: "0900"

   Please fix this one too (according to my previous remarks).

   —
   You are receiving this because you authored the thread.
   Reply to this email directly, view it on GitHub, or unsubscribe.

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+		capability "Configuration"
+		capability "Health Check"
+
+		fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName: "WiDom Dry Contact" // WiDom Dry Contact allow to turn On/Off a load with indipendent power supply.
```
(this remark was added previously, You might have missed it)
Please add deviceJoinName "WiDom Switch" in this fingerprint and add Raw Description in a comment after the fingerprint so it would look like:
fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName "WiDom Switch" // Raw Description (paste the raw description here)

(Raw description can be found at the Smartthings graphapi in device details)

GiuseppeWidom replied:
@KKlimczukS
Raw Description fixed on 1st April 2021

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+}
+
+private encap(physicalgraph.zwave.Command cmd) {
+	if (zwaveInfo.zw.contains("s")) {
```
Are both secEncap() and crcEncap() methods needed in this DTH (that supports one device) ?

GiuseppeWidom replied:
removed crcEncap

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,340 @@
+	}
+}
+
+private encapSequence(cmds, Integer delay=250) {
```
Could You merge those two encapSequence() methods into one that takes all the parameters (and set default value for delay and ep?

GiuseppeWidom replied:
Rmoved encapSequence

KKlimczukS commented:
@GiuseppeWidom any updates here ? (On Feb 11th I posted some new remarks)

GiuseppeWidom committed:
Fix on Raw description. Removed useless code.

KKlimczukS commented:
@GiuseppeWidom
Please post Your comments directly here, in this pull request: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/52683
Please, do not reply to github notifications. It will be more efficient to have this conversation via github comments under the remarks in this pull request.

GiuseppeWidom commented:
@KKlimczukS 
Any news?
have I to do other fix in the code?

PKacprowiczS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,325 @@
+		capability "Configuration"
+		capability "Health Check"
+
+		fingerprint mfr: "0149", prod: "1214", model: "0900", deviceJoinName: "WiDom Switch" // Raw Description WiDom Dry Contact allow to turn On/Off a load with indipendent power supply.
```
@GiuseppeWidom raw description is information describing z-wave device, it can be found in the IDE in device details section
![image](https://user-images.githubusercontent.com/41617389/113314963-16203700-930d-11eb-9413-92345da0fe14.png)

GiuseppeWidom replied:
Thanks @PKacprowiczS
I fixed Raw Description.
I made the code in this way in order to allow the end-user to use the Z-Wave configuration parameters of this device. The Z-Wave parameters can be used to change the device behaviour not only for automations or scenes. Every device has its Z-Wave configuration parameters. I think that remove the Z-Wave configuration parameters form a Z-Wave device is a pity. This is the first one (and more simple) of other devices that I'd like to add in Smartthings.

GiuseppeWidom replied:
@PKacprowiczS
I fixed the Raw description the 1st April 2021

PKacprowiczS reviewed (changes requested):
Honestly, I'm not sure how I feel about all that custom preferences. 
@GiuseppeWidom did you consider adding just a fingerprint to a generic DTH, like `zwave-switch`? Generally, we don't rely on z-wave scenes/associations, we have our own system for automations.

GiuseppeWidom committed:
Fix Raw Description

GiuseppeWidom commented:
Are there other changes that I have to do in order to obtain the merging?

KKlimczukS commented:
> Are there other changes that I have to do in order to obtain the merging?

@GiuseppeWidom Yes, there are still unresolved conversations in this pull request. Please answer to my previous questions.

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,325 @@
+	}
+}
+
+private secEncap(physicalgraph.zwave.Command cmd) {
```
Since this method is called only once, can You move that code to the encap() method ?

GiuseppeWidom replied:
Fixed as you requested.

KKlimczukS commented on devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy:
```diff
@@ -0,0 +1,325 @@
+	}
+}
+
+def on() {
```
Don't You need to poll the device for current state (BasicGet) or does the device report those changes itself ?
(comment applies to on() and off() methods)

GiuseppeWidom replied:
The device reports state changes itself.

GiuseppeWidom committed:
Moved secEncap method code inside the encap method
Fix on num clicks key control name.

lysst reviewed (commented):
Nice
